test/unit/uri.c

```diff
@@ -583,16 +590,195 @@ test_invalid_string_uri_set(void)
 	footer();
 }
 
+#define RFC3986_unreserved "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-._~"
+
+const struct str_escape escape_testcase[] = {
+	[0] = {
+		.str = "-._~",
+		.escaped = "-._~",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[1] = {
+		.str = "0123456789",
+		.escaped = "0123456789",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[2] = {
+		.str = "abcdefghijklm",
+		.escaped = "abcdefghijklm",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[3] = {
+		.str = "nopqrstuvwxyz",
+		.escaped = "nopqrstuvwxyz",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[4] = {
+		.str = "ABCDEFGHIJKLM",
+		.escaped = "ABCDEFGHIJKLM",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[5] = {
+		.str = "NOPQRSTUVWXYZ",
+		.escaped = "NOPQRSTUVWXYZ",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[6] = {
+		.str = "!$&'()*+,;=",
+		.escaped = "%21%24%26%27%28%29%2A%2B%2C%3B%3D",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+};
+
+/**
+ * Builds an array with unreserved characters.
+ * uri.unreserved() implemented as a Lua function,
+ * unreserved_tbl() replaces Lua implementation for testing purposes.
+ */
+static void
+unreserved_tbl(const char *str, unsigned char unreserved[256])
+{
+	for (int i = 0; i < 256; i++)
+		unreserved[i] = 0;
+
+	for (; *str; str++) {
+		unsigned char ch = (unsigned char)*str;
+		unreserved[ch] = 1;
+	}
+}
+
+static void
+test_escape(void)
+{
+	header();
+	plan(lengthof(escape_testcase) * 2);
+	unsigned char unreserved[256];
+	for (unsigned i = 0; i < lengthof(escape_testcase); i++) {
+		const char *unescaped = escape_testcase[i].str;
+		const char *escaped = escape_testcase[i].escaped;
+		bool plus = escape_testcase[i].plus;
+		char *dst = xcalloc(strlen(unescaped) * 3, sizeof("char"));
+		unreserved_tbl(escape_testcase[i].unreserved, unreserved);
+		size_t dst_size = uri_escape(unescaped, strlen(unescaped),
+					     dst, unreserved, plus);
+		isnt(dst_size, 0, "escaped string ('%s') length != 0", dst);
+		is(strcmp(escaped, dst), 0, "escape: '%s' == '%s'",
+		   escaped, dst);
+		free(dst);
+	}
+	check_plan();
+	footer();
+}
+
+static void
+test_unescape(void)
+{
+	header();
+	plan(lengthof(escape_testcase) * 2);
+	for (unsigned i = 0; i < lengthof(escape_testcase); i++) {
+		const char *unescaped = escape_testcase[i].str;
+		const char *escaped = escape_testcase[i].escaped;
+		bool decode_plus = escape_testcase[i].plus;
+		char *dst = xcalloc(strlen(unescaped), sizeof("char"));
+		size_t dst_size = uri_unescape(escaped, strlen(escaped),
+					       dst, decode_plus);
+		isnt(strlen(dst), 0, "unescaped string ('%s') length != 0",
+		     dst);
+		is(strcmp(dst, unescaped), 0, "unescape: '%s' == '%s'",
+		   unescaped, dst);
+		free(dst);
+	}
+	check_plan();
+	footer();
+}
```

Comment:
We're just lucky that `calloc()` allocates some memory after our `dst` buffer and it contains zeros. It is due to allocation granularity and, I guess, because `calloc()` uses it own zero filled slabs pool.

In fact, if a non-zero byte will appear afterwards, the test will fail. You can try it: extend the buffer on one byte and write a non-zero to this last byte.

```diff
diff --git a/test/unit/uri.c b/test/unit/uri.c
index bbb4e2b91..1f50e6aed 100644
--- a/test/unit/uri.c
+++ b/test/unit/uri.c
@@ -686,7 +686,8 @@ test_unescape(void)
                const char *unescaped = escape_testcase[i].str;
                const char *escaped = escape_testcase[i].escaped;
                bool decode_plus = escape_testcase[i].plus;
-               char *dst = xcalloc(strlen(unescaped), sizeof("char"));
+               char *dst = xcalloc(strlen(unescaped) + 1, sizeof("char"));
+               dst[strlen(unescaped)] = 'x';
                size_t dst_size = uri_unescape(escaped, strlen(escaped),
                                               dst, decode_plus);
                isnt(strlen(dst), 0, "unescaped string ('%s') length != 0",
```

OTOH, `xcalloc()` on `strlen(unescaped) + 1` should work, because our test cases have no zeros in a middle. So I propose to patch it this way:

```diff
diff --git a/test/unit/uri.c b/test/unit/uri.c
index bbb4e2b91..49e1af663 100644
--- a/test/unit/uri.c
+++ b/test/unit/uri.c
@@ -664,7 +664,7 @@ test_escape(void)
 		const char *unescaped = escape_testcase[i].str;
 		const char *escaped = escape_testcase[i].escaped;
 		bool plus = escape_testcase[i].plus;
-		char *dst = xcalloc(strlen(unescaped) * 3, sizeof("char"));
+		char *dst = xcalloc(strlen(unescaped) * 3 + 1, sizeof("char"));
 		unreserved_tbl(escape_testcase[i].unreserved, unreserved);
 		size_t dst_size = uri_escape(unescaped, strlen(unescaped),
 					     dst, unreserved, plus);
@@ -686,7 +686,7 @@ test_unescape(void)
 		const char *unescaped = escape_testcase[i].str;
 		const char *escaped = escape_testcase[i].escaped;
 		bool decode_plus = escape_testcase[i].plus;
-		char *dst = xcalloc(strlen(unescaped), sizeof("char"));
+		char *dst = xcalloc(strlen(unescaped) + 1, sizeof("char"));
 		size_t dst_size = uri_unescape(escaped, strlen(escaped),
 					       dst, decode_plus);
 		isnt(strlen(dst), 0, "unescaped string ('%s') length != 0",
@@ -753,7 +753,7 @@ test_unescape_special_cases(void)
 		const char *unescaped = unescape_testcase[i].str;
 		const char *escaped = unescape_testcase[i].escaped;
 		bool decode_plus = escape_testcase[i].plus;
-		char *dst = xcalloc(strlen(unescaped), sizeof("char"));
+		char *dst = xcalloc(strlen(unescaped) + 1, sizeof("char"));
 		size_t dst_size = uri_unescape(escaped, strlen(escaped),
 					       dst, decode_plus);
 		isnt(strlen(dst), 0, "unescaped string ('%s') length != 0",
```